In [ ]:
#r "nuget: FsODE, 0.0.1"
#r "nuget: FSharp.Stats, 0.4.7"
#r "nuget: Plotly.NET.Interactive, 3.0.2"

open FsODE
open Plotly.NET
open Plotly.NET.LayoutObjects
open Plotly.NET.StyleParam
open FSharp.Stats

Installed Packages FSharp.Stats, 0.4.7 FsODE, 0.0.1 Plotly.NET.Interactive, 3.0.2

# Tag 1 – Einführung in die Wachstumsmodellierung und Grundlagen einfacher Differentialgleichungen

Es gibt verschiedene Aspekte in biologischen  Systemen und Gesetzmäßigkeiten, die man modellieren kann. Ein besonders häufiger und beliebter Aspekt in der Biologie ist das Wachstum. Das Wachstum von Zellen, von Geweben, von Organen und Organismen, von Populationen und von Stoffmengen. Dazu gehört auch das negative Wachstum, auch Zerfall oder Abfall genannt. Bspw. der Verbrauch von Substrat durch Lebewesen in einer Umgebung oder das Eingehen (Absterben) von Pflanzenindividuen auf einer Wiese infolge von anhaltender Dürre und Trockenheit.

Am ersten Tag des Grundpraktikums "Biophysik" behandeln wir grundlegende Dinge über Wachstum und deren Modellierung.

## Was ist Wachstumsmodellierung?

Wissenschaftliche Modelle sind Abstraktionen der Wirklichkeit. Das Ziel ist, mit einem Modell einen (meist komplexen) Sachverhalt zu beschreiben.  
Um Wachstum zu beschreiben, verwenden wir gedankliche Modelle (_mental models_), die biologische Gesetzmäßigkeiten möglichst präzise abbilden. Modellierung an sich ist in der Biologie sehr wichtig, um Veränderungen und Wechselwirkungen erklären und mehr oder weniger präzise (je nach Güte des Modells) vorhersagen zu können.

Zu den ältesten und _vermeintlich_ einfachsten Experimenten gehört die Analyse von Wachstum. Dieses kann je nach untersuchtem System durch verschiedene Parameter beschrieben werden. Steht der Organismus im Fokus, so sind Gewicht, Länge oder Größe von Interesse. Bei Populationen geht es meist um eine Zell- oder Individuenzahl pro definiertem Volumen oder Fläche (Populationsdichte).
Im Labor können verschiedene Wachstumsbedingungen künstlich erzeugt werden und damit der Einfluss jedes einzelnen Faktors – getrennt oder kombiniert – untersucht werden. Zum Beispiel können Bakterien statt auf Standardmedium mit Glucose auf Medium mit Galactose angezogen werden. Eine eventuell nötige Umwandlung von Galactose in Glucose, oder aufwendigere metabolische Reaktionen führen hier zu langsamerem Wachstum. Dies lässt sich in der Regel durch einfache Absorptionsmessungen im Photometer beweisen. Ein vermindertes Wachstum geht hier mit einer langsameren Trübung (niedrigere optische Dichte) des Mediums einher, was auf einem Graphen leicht visualisiert werden kann.

<center>
<img src="../../images/day1/02_Growth_01.png" width=40% /> <br>
<i><b>Abbildung 1: Beispiel für einen Wachstumsverlauf.</b> Aufzucht von Bakterien in einem Glucose- bzw. Galactose-haltigen Medium. Die y-Achse gibt die Absorption im Photometer bei einer bestimmten optischen Dichte an.</i>
</center>

Leider kann der Wachstumsunterschied mit dem oben gezeigten Datensatz nicht quantifiziert werden. Zwar kann die zeitliche Differenz zwischen zwei Punkten gleicher optischer Dichte bestimmt werden (Kreuze in Abbildung 1), jedoch ist diese Differenz sowohl stark von den gewählten Messzeitpunkten, als auch von den vorliegenden (unbekannten) Wachstumsphasen abhängig. Außerdem verläuft die Trübung nicht linear. Das bedeutet, dass sich bei einer Zellverdopplung nicht automatisch auch die Absorption verdoppelt. Für qualitative Aussagen ist diese Methode ausreichend, für quantitative Analysen aber ungeeignet.

Wesentlich aussagekräftiger ist eine Bestimmung der absoluten Zellkonzentration mit anschließendem Fitting eines bekannten Wachstumsmodells. Die so ermittelten Parameter können für quantitative und vergleichende Analysen von z.B. Glucose- und Galactosemedium verwendet werden. Das einfachste Wachstumsmodell wollen wir uns im Folgenden anschauen.  

____

## Exkurs: Differentialgleichungen

Biologische Prozesse können meist mit Differentialgleichungen (DGLs) dargestellt werden. Bei einer Differentialgleichung befinden sich sowohl die Funktion als auch eine (oder mehrere) ihrer Ableitungen in einer Gleichung. Die exakte Lösung einer solchen DGL ergibt die Funktion, mit der sich der biologische Sachverhalt (hier: die Wachstumsfunktion) adäquat beschreiben lässt.

$\frac{df}{dx} = 4f(x)$, eine beispielhafte DGL

Falls Sie Ihr Schulwissen über Gleichungen, Funktionen und Ableitungen auffrischen möchten, lesen Sie [dieses Notebook](link einfügen).

___

## Einfaches Wachstumsmodell

Wachstum, also die Änderung der Zellzahl über die Zeit, ist proportional zur Zellzahl _N_. Der Proportionalitätsfaktor ist _r_. Je höher die Zellzahl, desto höher ist die Rate mit der die Zellzahl wächst (Diese Begründung bitte kurz sacken lassen).

_Aufgabe 1.1:_ Wie sieht die beschriebene Differentialgleichung aus?

a) $\frac{dN}{dt} = \frac{r}{N}$

b) $N(t) = rtN$

c) $N(t) = \frac{rN}{t}$

d) $\frac{dN}{dt} = rN$


_Antwort:_ 

_Aufgabe 1.2:_ Beweisen Sie, dass $N(t) = N_0 *e^{rt}$ eine Lösung für die oben definierte Differentialgleichung ist.

Hinweis: Leiten Sie die oben stehende geschlossene Form der Wachstumsgleichung nach $t$ ab und setzen Sie $N$ und die Ableitung in die Differenzialgleichung ein. Versuchen Sie diese Aufgabe innerhalb von 10 Minuten zu lösen. Fahren Sie ansonsten mit dem nächsten Abschnitt fort.

## Lösen von DGLs

Nicht immer ist eine Integration von Modellen möglich. Wie oben bewiesen ist das Integral tatsächlich eine Lösung der DGL. 
In der Biologie gibt es oftmals Fragestellungen oder Systeme, in denen eine Vielzahl von Parametern und Abhängigkeiten eine Rolle spielen und somit eine Integration nicht machbar ist. Glücklicherweise ist dies auch nicht notwendig. Denn das Wissen um das Verhalten der Steigung beschreibt den Funktionsverlauf meist ausreichend.  
In der Mathematik erarbeitet man die Lösung einer solchen DGL über verschiedene (mögliche) Lösungswege, jedoch sind für viele DGLs nur approximative (= Näherungs-)Lösungen möglich.

### Euler-Verfahren

Ein Beispiel für eine solche Methode ist das sogenannte **Euler-Verfahren**. Es handelt sich dabei um einen iterativen Algorithmus, d. h., dass eine bestimmte Berechnung mehrmals wiederholt wird, bei jeder Wiederholung jedoch mit veränderten Startbedingungen.  
Für das Euler-Verfahren wird mind. 1 Startpunkt benötigt, z. B. $P(0|1)$. Daher wissen wir, dass der Kurvenverlauf auf jeden Fall an der x-Stelle $0$ $y = 1$ sein muss. Das bedeutet:

$N(0) = 1$

Zudem setzen wir eine sog. Schrittweite (wir nennen sie $h$). Die sagt aus, wie weit wir bei jedem Iterationsschritt nach rechts gehen (also um wieviel sich der $x$-Wert erhöht). In unserem Beispiel sei das:

$h = 1$

Jetzt 

$\frac{dN}{dt} = rN$

_Aufgabe 1.3.:_ Wie könnten Sie als Biolog:in einen solchen Startpunkt P ermitteln?

_Antwort:_

Auf dem Papier würde eine solche Rechnung lange dauern. In der Informatik benutzt man daher eine Programmbibliothek, mit deren Hilfe wir uns vom Computer die Lösung _errechnen lassen_. Das einzige, was wir noch tun müssen, ist, unserem "Taschenrechner" die korrekte DGL zu geben.

Wieder unsere DGL von oben: $\frac{\mathrm{d}N}{\mathrm{d}t} = rN$

Nachdem das Wissen über Ableitungen aufgefrischt wurden, soll die Wachstumsfunktion nun simuliert werden. Dafür wird die Differenzialgleichung als Modell formalisiert und mit folgenden Werten parametrisiert:

$$r = 0.3$$

Damit der Algorithmus nun die gesuchte Funktion errechnen kann, braucht er er Anfangswerte.
Wir geben ihm die Anfangswerte  

$$N(t0) = 100$$

$$t0 = 0.0$$

In [ ]:
// der Rechenalgorithmus der Programmbibliothek, die wir benutzen:
let modelContext = 
    OdeContext(
        OdeSolverMethod.RK546M,
        OdeSolverOptions(
            StepSize=0.42
        )
    )
    
// unsere DGL als einfaches Modell (SimpleModel):
let r = 0.3  

// growth model
let dN_dt : SimpleModel =     
    fun N t ->
        let cellcount = N
        let cellcount' = r * N 
        cellcount'    

// Initial condition
let t0 = 0         
let N0 = 100.0    


FSI_0009+it@24

Im Modell dN_dt sind nun alle relevanten Berechnungen bereits geschehen und man kann sich die Funktionswerte der simulierten Integration ausgeben lassen.

In [ ]:
//xxxx

let sim_dy_dx = 
    modelContext.OdeInt(t0,N0,dN_dt) //aufgabe
    |> SolPoints.take 10 
    |> SolPoints.memorize

sim_dy_dx

index x Y 0 0 [ 100 ] 1 0.42 [ 113.4282167625625 ] 2 0.999259549657388 [ 134.955898363052 ] 3 2.1772271605451294 [ 192.16186029123628 ] 4 3.8940775510006906 [ 321.6263247377607 ] 5 6.039496988444697 [ 612.1674466254168 ] 6 8.515955468393773 [ 1286.8070827545662 ] 7 11.251389834936289 [ 2923.4116361761935 ] 8 14.195179461744015 [ 7069.730547285138 ] 9 17.312471927868035 [ 18009.84524032065 ]

_Aufgabe 2.4: Visualisieren Sie das Modell anhand der oben zwischengespeicherten 10 Datenpunkte und beschriften Sie die x- und y-Achse._

In [ ]:

let shapeOfLine =
    LayoutObjects.Shape.init(ShapeType.Line,sim_dy_dx.[1].x,sim_dy_dx.[3].x,sim_dy_dx.[1].Y.[0],sim_dy_dx.[3].Y.[0],Line=Line.init(Dash=DrawingStyle.Dash))


// Plot the numerical solution
sim_dy_dx
|> SolPoints.toPoints 1
|> Chart.Point
|> Chart.withShape(shapeOfLine)
|> Chart.withTitle("Abbildung 2 - Simulation simple growth model")
|> Chart.withXAxisStyle("Time") //aufgabe
|> Chart.withYAxisStyle("Population size") //aufgabe

<!-- Plotly chart will be drawn inside this DIV -->

_Aufgabe 2.6: Schätzen Sie grob die Steigung zum Zeitpunkt t=1 ab. Entpricht Sie der theoretischen Steigung aus der Differentialgleichung (m=rN)?_
  - Berechnen Sie die Steigung wie im unteren Schema gezeigt
    - Zoomen und hovern Sie über die Punkte um die Werte bei t=0.42 und t=2.18 angezeigt zu bekommen.
  - Berechnen Sie die Steigung anhand der Differentialgleichung
    - Bestimmen Sie N an t=1
    - r ist gegeben mit r=0.3
    - Berechnen Sie die Steigung m=rN
  - Vergleichen Sie die Steigung durch Abschätzen mit der theoretischen Steigung der Differentialgleichung. Sind sie ähnlich?

<center>
<img src="../../images/day1/02_Growth_slope.png" width=30% />
</center>

_Aufgabe 2.7: Formulieren Sie die Wachstumsfunktion aus Aufgabe 2.2._

In [ ]:
let y_exact =
    fun t -> 
        N0*exp(r*t) //aufgabe

[
    sim_dy_dx
    |> SolPoints.toPoints 1
    |> Chart.Spline 
    |> Chart.withTraceInfo("simulation");
    
    sim_dy_dx
    |> SolPoints.map (fun p -> (p.x, y_exact p.x))
    |> Chart.Point
    |> Chart.withTraceInfo("exact");

]
|> Chart.combine
|> Chart.withTitle("Abbildung 3 - Simple growth model comparison")
|> Chart.withXAxisStyle("Time")
|> Chart.withYAxisStyle("Population size")

Die Simulation der Differenzialgleichung entspricht der exakten Lösung der geschlossenen Form.



Es folgt eine kurze Besprechung. Falls Sie schnell vorangekommen sind, können sie bis zur Besprechung folgende Aufgaben lösen:


### Expertenaufgaben
- Integrieren Sie $\frac{dN}{dt}=rN$ zu $N(t) = N_0 *e^{rt}$

___



## Aufgaben

*Aufgabe 1.1:*



*Antwort:* 

*Aufgabe 1.2:* 

*Antwort:* 

*Aufgabe 2:* 

*Aufgabe 10:*

